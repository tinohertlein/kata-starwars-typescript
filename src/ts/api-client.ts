const BASE_URL = 'https://swapi.dev/api'

export interface Character {
    name: string
    homeworldUrl: string
}

export interface Planet {
    name: string
    url: string
}

export const fetchCharacters = async () => fetchEntities('people', toCharacters)

export const fetchPlanets = async () => fetchEntities('planets', toPlanets)

const fetchEntities = async (path: string, toEntity: (json: any) => any[]) => {
    const firstPageUrl = `${BASE_URL}/${path}`
    const entities = [] as any[]

    let pageUrl = firstPageUrl

    while (pageUrl) {
        const json = await fetchPage(pageUrl)
        pageUrl = json.next
        entities.push(...toEntity(json))
    }

    return entities
}

const fetchPage = async (url: string) => await (await fetch(url)).json()

const toCharacters = (json: any) =>
    json.results.map(
        (result: any) =>
            ({
                name: result.name,
                homeworldUrl: result.homeworld,
            } as Character)
    )

const toPlanets = (json: any) =>
    json.results.map(
        (result: any) =>
            ({
                name: result.name,
                url: result.url,
            } as Planet)
    )
