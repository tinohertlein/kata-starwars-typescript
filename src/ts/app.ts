import { fetchCharacters, fetchPlanets } from './api-client'
import jquery from 'jquery'

export const populateTableBody = () =>
    Promise.all([fetchCharacters(), fetchPlanets()])
        .then((entities) => {
            const [characters, planets] = entities

            const tbody = jquery('tbody').empty()

            characters.forEach((character) => {
                const homeworldName = planets.find(
                    (planet) => planet.url === character.homeworldUrl
                )?.name

                const tr = jquery('<tr></tr>')
                const tdCharacter = jquery(`<td>${character.name}</td>`)
                const tdHomeWorld = jquery(`<td>${homeworldName}</td>`)
                tr.append(tdCharacter)
                tr.append(tdHomeWorld)
                tbody.append(tr)
            })
        })
        .catch((error) => console.error(error))
