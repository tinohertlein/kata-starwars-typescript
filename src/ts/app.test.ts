/**
 * @jest-environment jsdom
 */
import * as apiClient from './api-client'
import {Character, Planet} from './api-client'
import {populateTableBody} from './app'
import jquery from 'jquery'
import Mock = jest.Mock

describe('populateTable should', () => {
    console.error = jest.fn()

    beforeEach(() => {
        ;(console.error as Mock).mockClear()
    })

    test('display the characters & their homeworlds', async () => {
        const tbody = jquery('body').append('<tbody></tbody>')
        const fetchCharactersSpy = jest.spyOn(apiClient, 'fetchCharacters').mockResolvedValue([
            {
                name: 'Leia',
                homeworldUrl: 'tatooine-url',
            } as Character,
            {
                name: 'Luke',
                homeworldUrl: 'alderaan-url',
            } as Character
        ])
        const fetchPlanetsSpy = jest.spyOn(apiClient, 'fetchPlanets').mockResolvedValue([
            {
                name: 'Tatooine',
                url: 'tatooine-url',
            } as Planet,
            {
                name: 'Alderaan',
                url: 'alderaan-url',
            } as Planet
        ])

        await populateTableBody()

        expect(fetchCharactersSpy).toBeCalled()
        expect(fetchPlanetsSpy).toBeCalled()
        expect(tbody.get()).toMatchSnapshot()
    })

    test('write errors to the console', async () => {
        const errorMessage = 'This is an error';
        jest.spyOn(apiClient, 'fetchCharacters').mockRejectedValue(
            errorMessage
        )

        await populateTableBody()

        expect(console.error).toHaveBeenCalledWith(errorMessage)
    })
})
