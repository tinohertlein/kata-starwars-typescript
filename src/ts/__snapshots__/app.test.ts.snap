// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`populateTable should display the characters & their homeworlds 1`] = `
Array [
  <body>
    <tbody>
      <tr>
        <td>
          Leia
        </td>
        <td>
          Tatooine
        </td>
      </tr>
      <tr>
        <td>
          Luke
        </td>
        <td>
          Alderaan
        </td>
      </tr>
    </tbody>
  </body>,
]
`;
